pkg/controller/direct/apigee/instance_controller.go

```diff
@@ -0,0 +1,264 @@
+		log.V(2).Info("change detected: consumerAcceptList")
+		updateMask.Paths = append(updateMask.Paths, "consumerAcceptList")
+	}
+
+	if len(updateMask.Paths) == 0 {
+		log.V(2).Info("no field needs update", "name", a.id)
+		status := &krm.ApigeeInstanceStatus{}
+		status.ObservedState = ApigeeInstanceObservedState_FromAPI(mapCtx, a.actual)
+		if mapCtx.Err() != nil {
+			return mapCtx.Err()
+		}
+		return updateOp.UpdateStatus(ctx, status, nil)
+	}
+
+	op, err := a.instancesClient.Patch(a.id.String(), resource).UpdateMask(strings.Join(updateMask.Paths, ",")).Context(ctx).Do()
+	if err != nil {
+		return fmt.Errorf("updating ApigeeInstance %s: %w", a.id, err)
+	}
+	if err := a.waitForOp(ctx, op); err != nil {
+		return fmt.Errorf("ApigeeInstance %s waiting update: %w", a.id, err)
+	}
+	updated, err := a.instancesClient.Get(a.id.String()).Context(ctx).Do()
+	if err != nil {
+		return fmt.Errorf("getting updated ApigeeInstance: %w", err)
+	}
+	log.V(2).Info("successfully updated ApigeeInstance", "name", a.id)
+
+	status := &krm.ApigeeInstanceStatus{}
+	status.ObservedState = ApigeeInstanceObservedState_FromAPI(mapCtx, updated)
+	if mapCtx.Err() != nil {
+		return mapCtx.Err()
+	}
+	return updateOp.UpdateStatus(ctx, status, nil)
+}
+
+// Export maps the GCP object to a Config Connector resource `spec`.
+func (a *ApigeeInstanceAdapter) Export(ctx context.Context) (*unstructured.Unstructured, error) {
+	if a.actual == nil {
+		return nil, fmt.Errorf("Find() not called")
+	}
+	u := &unstructured.Unstructured{}
+
+	obj := &krm.ApigeeInstance{}
+	mapCtx := &direct.MapContext{}
+	obj.Spec = direct.ValueOf(ApigeeInstanceSpec_FromAPI(mapCtx, a.actual))
+	if mapCtx.Err() != nil {
+		return nil, mapCtx.Err()
+	}
+	obj.Spec.OrganizationRef = &refs.ApigeeOrganizationRef{External: a.id.Parent().String()}
+	uObj, err := runtime.DefaultUnstructuredConverter.ToUnstructured(obj)
+	if err != nil {
+		return nil, err
+	}
+
+	u.SetName(a.id.ID())
+	u.SetGroupVersionKind(krm.ApigeeInstanceGVK)
+
+	u.Object = uObj
+	return u, nil
+}
+
+// Delete the resource from GCP service when the corresponding Config Connector resource is deleted.
+func (a *ApigeeInstanceAdapter) Delete(ctx context.Context, deleteOp *directbase.DeleteOperation) (bool, error) {
+	log := klog.FromContext(ctx)
+	log.V(2).Info("deleting ApigeeInstance", "name", a.id)
+
+	op, err := a.instancesClient.Delete(a.id.String()).Context(ctx).Do()
+	if err != nil {
+		if direct.IsNotFound(err) {
+			// Return success if not found (assume it was already deleted)
+			return true, nil
+		}
+		return false, fmt.Errorf("deleting ApigeeInstance %s: %w", a.id, err)
+	}
+	log.V(2).Info("successfully deleted ApigeeInstance", "name", a.id)
+
+	if err := a.waitForOp(ctx, op); err != nil {
+		return false, fmt.Errorf("waiting delete ApigeeInstance %s: %w", a.id, err)
+	}
+	return true, nil
+}
+
+func (a *ApigeeInstanceAdapter) waitForOp(ctx context.Context, op *api.GoogleLongrunningOperation) error {
```

Comment:
This can be an individual function. And suggest moving it to the controller/direct/common lib for sharing purpose.